File: android/src/main/kotlin/io/github/stanleycocos/flutter_system_events/FlutterSystemEventsPlugin.kt
package io.github.stanleycocos.flutter_system_events

import android.app.Activity
import android.app.Application
import android.content.BroadcastReceiver
import android.content.ComponentCallbacks2
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.res.Configuration
import android.database.ContentObserver
import android.graphics.Rect
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.net.Uri
import android.os.BatteryManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import android.util.Log
import android.view.Surface
import android.view.ViewTreeObserver
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlutterSystemEventsPlugin */
class FlutterSystemEventsPlugin :
    FlutterPlugin,
    MethodCallHandler,
    EventChannel.StreamHandler,
    ActivityAware {
    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var appContext: Context? = null
    private var events: EventChannel.EventSink? = null
    private var activity: Activity? = null
    private var initialized = false
    private var lifecycleCallbacks: Application.ActivityLifecycleCallbacks? = null
    private var keyboardListener: ViewTreeObserver.OnGlobalLayoutListener? = null
    private var keyboardVisible = false
    private var networkCallback: ConnectivityManager.NetworkCallback? = null
    private var memoryCallbacks: ComponentCallbacks2? = null
    private var batteryReceiver: BroadcastReceiver? = null
    private var timeReceiver: BroadcastReceiver? = null
    private var screenReceiver: BroadcastReceiver? = null
    private var brightnessObserver: ContentObserver? = null
    private var orientationCallbacks: ComponentCallbacks2? = null
    private var lastOrientation: String? = null
    private var config = EventConfig.legacy()
    private val mainHandler by lazy { Handler(Looper.getMainLooper()) }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        appContext = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_system_events")
        channel.setMethodCallHandler(this)
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "flutter_system_events/events")
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result
    ) {
        when (call.method) {
            "initialize" -> {
                config = EventConfig.from(call.arguments)
                initialized = true
                startAll()
                result.success(null)
            }
            "dispose" -> {
                initialized = false
                stopAll()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        stopAll()
        initialized = false
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        appContext = null
    }

    override fun onListen(arguments: Any?, eventSink: EventChannel.EventSink) {
        events = eventSink
        if (initialized) startAll()
    }

    override fun onCancel(arguments: Any?) {
        events = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        if (initialized) startAll()
    }

    override fun onDetachedFromActivityForConfigChanges() {
        stopAll()
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        if (initialized) startAll()
    }

    override fun onDetachedFromActivity() {
        stopAll()
        activity = null
    }

    private fun startAll() {
        stopAll()
        if (config.keyboard) startKeyboard()
        if (config.lifecycle) startLifecycle()
        if (config.network) startNetwork()
        if (config.memory) startMemory()
        if (config.battery) startBattery()
        if (config.orientation) startOrientation()
        if (config.time) startTime()
        if (config.screen) startScreen()
    }

    private fun stopAll() {
        stopKeyboard()
        stopLifecycle()
        stopNetwork()
        stopMemory()
        stopBattery()
        stopOrientation()
        stopTime()
        stopScreen()
    }

    private fun emitEvent(event: Map<String, Any>) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            events?.success(event)
        } else {
            mainHandler.post { events?.success(event) }
        }
    }

    private fun startLifecycle() {
        val currentActivity = activity ?: return
        val callbacks = object : Application.ActivityLifecycleCallbacks {
            override fun onActivityResumed(activity: Activity) = emitLifecycle(activity, "resumed")
            override fun onActivityPaused(activity: Activity) = emitLifecycle(activity, "inactive")
            override fun onActivityStopped(activity: Activity) = emitLifecycle(activity, "paused")
            override fun onActivityDestroyed(activity: Activity) = emitLifecycle(activity, "detached")
            override fun onActivityStarted(activity: Activity) {}
            override fun onActivityCreated(activity: Activity, state: Bundle?) {}
            override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
        }
        currentActivity.application.registerActivityLifecycleCallbacks(callbacks)
        lifecycleCallbacks = callbacks
    }

    private fun emitLifecycle(source: Activity, state: String) {
        if (source == activity) emitEvent(mapOf("type" to "lifecycle", "state" to state))
    }

    private fun stopLifecycle() {
        val currentActivity = activity
        lifecycleCallbacks?.let { currentActivity?.application?.unregisterActivityLifecycleCallbacks(it) }
        lifecycleCallbacks = null
    }

    private fun startKeyboard() {
        val root = activity?.window?.decorView ?: return
        val listener = ViewTreeObserver.OnGlobalLayoutListener {
            val rect = Rect()
            root.getWindowVisibleDisplayFrame(rect)
            val height = root.rootView.height
            val keyboardHeight = (height - rect.bottom).coerceAtLeast(0)
            val visible = keyboardHeight > height * 0.15
            if (visible != keyboardVisible) {
                keyboardVisible = visible
                val logicalKeyboardHeight = logicalPixelsFromPhysicalPixels(
                    keyboardHeight,
                    root.resources.displayMetrics.density,
                )
                emitEvent(
                    mapOf(
                        "type" to "keyboard",
                        "visible" to visible,
                        "height" to if (visible) logicalKeyboardHeight else 0,
                    ),
                )
            }
        }
        root.viewTreeObserver.addOnGlobalLayoutListener(listener)
        keyboardListener = listener
    }

    private fun stopKeyboard() {
        val root = activity?.window?.decorView
        keyboardListener?.let { root?.viewTreeObserver?.removeOnGlobalLayoutListener(it) }
        keyboardListener = null
        keyboardVisible = false
    }

    private fun startNetwork() {
        val manager = appContext?.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager ?: return
        val callback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) = emitNetwork(manager)
            override fun onLost(network: Network) = emitNetwork(manager)
            override fun onCapabilitiesChanged(network: Network, capabilities: NetworkCapabilities) = emitNetwork(manager)
        }
        manager.registerNetworkCallback(NetworkRequest.Builder().build(), callback)
        networkCallback = callback
        emitNetwork(manager)
    }

    private fun emitNetwork(manager: ConnectivityManager) {
        val capabilities = manager.getNetworkCapabilities(manager.activeNetwork)
        val online = capabilities?.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) == true
        val networkType = when {
            !online -> "none"
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> "wifi"
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> "cellular"
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> "ethernet"
            else -> "other"
        }
        emitEvent(mapOf("type" to "network", "online" to online, "networkType" to networkType))
    }

    private fun stopNetwork() {
        val manager = appContext?.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
        networkCallback?.let { manager?.unregisterNetworkCallback(it) }
        networkCallback = null
    }

    private fun startMemory() {
        val context = appContext ?: return
        val callbacks = object : ComponentCallbacks2 {
            override fun onLowMemory() = emitMemory("low", 0)
            override fun onTrimMemory(level: Int) = emitMemory("trim", level)
            override fun onConfigurationChanged(newConfig: Configuration) {}
        }
        context.registerComponentCallbacks(callbacks)
        memoryCallbacks = callbacks
    }

    private fun emitMemory(state: String, level: Int) {
        Log.d("FlutterSystemEvents", "memory callback: state=$state level=$level")
        emitEvent(mapOf("type" to "memory", "state" to state, "level" to level))
    }

    private fun stopMemory() {
        memoryCallbacks?.let { appContext?.unregisterComponentCallbacks(it) }
        memoryCallbacks = null
    }

    private fun startBattery() {
        val context = appContext ?: return
        val filter = IntentFilter(Intent.ACTION_BATTERY_CHANGED)
        val receiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) = emitBattery(intent)
        }
        batteryReceiver = receiver
        context.registerReceiver(receiver, filter)
        context.registerReceiver(null, filter)?.let(::emitBattery)
    }

    private fun emitBattery(intent: Intent) {
        val level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
        val scale = intent.getIntExtra(BatteryManager.EXTRA_SCALE, -1)
        val status = intent.getIntExtra(BatteryManager.EXTRA_STATUS, BatteryManager.BATTERY_STATUS_UNKNOWN)
        val percent = if (level >= 0 && scale > 0) level * 100 / scale else -1
        val state = when (status) {
            BatteryManager.BATTERY_STATUS_CHARGING -> "charging"
            BatteryManager.BATTERY_STATUS_DISCHARGING,
            BatteryManager.BATTERY_STATUS_NOT_CHARGING -> "discharging"
            BatteryManager.BATTERY_STATUS_FULL -> "full"
            else -> "unknown"
        }
        emitEvent(
            mapOf(
                "type" to "battery",
                "level" to percent,
                "charging" to (state == "charging" || state == "full"),
                "state" to state,
            ),
        )
    }

    private fun stopBattery() {
        batteryReceiver?.let { appContext?.unregisterReceiver(it) }
        batteryReceiver = null
    }

    private fun startTime() {
        val context = appContext ?: return
        val filter = IntentFilter().apply {
            addAction(Intent.ACTION_TIME_CHANGED)
            addAction(Intent.ACTION_TIMEZONE_CHANGED)
            addAction(Intent.ACTION_DATE_CHANGED)
        }
        val receiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                emitEvent(mapOf("type" to "time", "reason" to timeReasonFromAction(intent.action)))
            }
        }
        timeReceiver = receiver
        context.registerReceiver(receiver, filter)
    }

    private fun stopTime() {
        timeReceiver?.let { appContext?.unregisterReceiver(it) }
        timeReceiver = null
    }

    private fun startScreen() {
        val context = appContext ?: return
        val filter = IntentFilter(Intent.ACTION_SCREEN_OFF).apply {
            addAction(Intent.ACTION_SCREEN_ON)
            addAction(Intent.ACTION_USER_PRESENT)
        }
        val receiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                emitEvent(screenEventFromAction(intent.action))
            }
        }
        screenReceiver = receiver
        context.registerReceiver(receiver, filter)

        val observer = object : ContentObserver(mainHandler) {
            override fun onChange(selfChange: Boolean, uri: Uri?) {
                emitBrightness(context)
            }
        }
        brightnessObserver = observer
        context.contentResolver.registerContentObserver(
            Settings.System.getUriFor(Settings.System.SCREEN_BRIGHTNESS),
            false,
            observer,
        )
    }

    private fun stopScreen() {
        screenReceiver?.let { appContext?.unregisterReceiver(it) }
        screenReceiver = null
        brightnessObserver?.let { appContext?.contentResolver?.unregisterContentObserver(it) }
        brightnessObserver = null
    }

    private fun emitBrightness(context: Context) {
        val raw = Settings.System.getInt(context.contentResolver, Settings.System.SCREEN_BRIGHTNESS, -1)
        emitEvent(screenBrightnessEvent(raw) ?: return)
    }

    private fun startOrientation() {
        val context = appContext ?: return
        val callbacks = object : ComponentCallbacks2 {
            override fun onConfigurationChanged(newConfig: Configuration) = emitOrientation()
            override fun onLowMemory() {}
            override fun onTrimMemory(level: Int) {}
        }
        context.registerComponentCallbacks(callbacks)
        orientationCallbacks = callbacks
        emitOrientation()
    }

    private fun emitOrientation() {
        val rotation = activity?.windowManager?.defaultDisplay?.rotation
        val orientation = orientationNameFromRotation(rotation)
        if (orientation == lastOrientation) return
        lastOrientation = orientation
        emitEvent(mapOf("type" to "orientation", "orientation" to orientation))
    }

    private fun stopOrientation() {
        orientationCallbacks?.let { appContext?.unregisterComponentCallbacks(it) }
        orientationCallbacks = null
        lastOrientation = null
    }

    private data class EventConfig(
        val keyboard: Boolean,
        val lifecycle: Boolean,
        val network: Boolean,
        val memory: Boolean,
        val battery: Boolean,
        val orientation: Boolean,
        val time: Boolean,
        val screen: Boolean,
    ) {
        companion object {
            fun legacy() = EventConfig(
                keyboard = true,
                lifecycle = true,
                network = true,
                memory = true,
                battery = false,
                orientation = true,
                time = true,
                screen = true,
            )

            fun from(arguments: Any?): EventConfig {
                val map = arguments as? Map<*, *> ?: return legacy()
                return EventConfig(
                    keyboard = map["keyboard"] == true,
                    lifecycle = map["lifecycle"] == true,
                    network = map["network"] == true,
                    memory = map["memory"] == true,
                    battery = map["battery"] == true,
                    orientation = map["orientation"] == true,
                    time = map["time"] == true,
                    screen = map["screen"] == true,
                )
            }
        }
    }
}

internal fun orientationNameFromRotation(rotation: Int?): String = when (rotation) {
    Surface.ROTATION_0 -> "portraitUp"
    Surface.ROTATION_90 -> "landscapeLeft"
    Surface.ROTATION_180 -> "portraitDown"
    Surface.ROTATION_270 -> "landscapeRight"
    else -> "unknown"
}

internal fun timeReasonFromAction(action: String?): String = when (action) {
    Intent.ACTION_TIME_CHANGED -> "timeChanged"
    Intent.ACTION_TIMEZONE_CHANGED -> "timezoneChanged"
    Intent.ACTION_DATE_CHANGED -> "dateChanged"
    else -> "unknown"
}

internal fun screenChangeFromAction(action: String?): String = when (action) {
    Intent.ACTION_SCREEN_OFF -> "off"
    Intent.ACTION_SCREEN_ON -> "on"
    Intent.ACTION_USER_PRESENT -> "unlocked"
    else -> "unknown"
}

internal fun screenEventFromAction(action: String?): Map<String, Any> =
    mapOf("type" to "screen", "change" to screenChangeFromAction(action))

internal fun screenBrightnessEvent(value: Int): Map<String, Any>? {
    val brightness = normalizedBrightness(value) ?: return null
    return mapOf("type" to "screen", "change" to "brightness", "brightness" to brightness)
}

internal fun normalizedBrightness(value: Int): Double? {
    if (value < 0) return null
    return value.coerceAtMost(255) / 255.0
}

internal fun logicalPixelsFromPhysicalPixels(pixels: Int, density: Float): Double =
    pixels / if (density > 0f) density.toDouble() else 1.0

File: ios/flutter_system_events/Sources/flutter_system_events/FlutterSystemEventsPlugin.swift
import Flutter
import Network
import UIKit

public class FlutterSystemEventsPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private var events: FlutterEventSink?
  private var observers: [NSObjectProtocol] = []
  private var pathMonitor: NWPathMonitor?
  private var currentNetworkMonitor: NWPathMonitor?
  private var config = EventConfig.legacy
  private var previousBatteryMonitoring: Bool?
  private var previousOrientationNotifications: Bool?
  private var lastOrientation: String?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_system_events", binaryMessenger: registrar.messenger())
    let eventChannel = FlutterEventChannel(name: "flutter_system_events/events", binaryMessenger: registrar.messenger())
    let instance = FlutterSystemEventsPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
    eventChannel.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initialize":
      config = EventConfig.from(call.arguments)
      startAll()
      result(nil)
    case "dispose":
      stopAll()
      result(nil)
    case "currentNetwork":
      currentNetwork(result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    self.events = events
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    events = nil
    return nil
  }

  private func startAll() {
    stopAll()
    if config.keyboard { startKeyboard() }
    if config.lifecycle { startLifecycle() }
    if config.network { startNetwork() }
    if config.memory { startMemory() }
    if config.battery { startBattery() }
    if config.orientation { startOrientation() }
    if config.time { startTime() }
    if config.screen { startScreen() }
  }

  private func startKeyboard() {
    observers.append(NotificationCenter.default.addObserver(forName: UIResponder.keyboardWillShowNotification, object: nil, queue: .main) { [weak self] notification in
      let frame = notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect
      self?.events?(["type": "keyboard", "visible": true, "height": frame?.height ?? 0])
    })
    observers.append(NotificationCenter.default.addObserver(forName: UIResponder.keyboardWillHideNotification, object: nil, queue: .main) { [weak self] _ in
      self?.events?(["type": "keyboard", "visible": false, "height": 0])
    })
  }

  private func startLifecycle() {
    observeLifecycle(UIApplication.didBecomeActiveNotification, state: "resumed")
    observeLifecycle(UIApplication.willResignActiveNotification, state: "inactive")
    observeLifecycle(UIApplication.didEnterBackgroundNotification, state: "paused")
    observeLifecycle(UIApplication.willTerminateNotification, state: "detached")
  }

  private func observeLifecycle(_ name: Notification.Name, state: String) {
    observers.append(NotificationCenter.default.addObserver(forName: name, object: nil, queue: .main) { [weak self] _ in
      self?.events?(["type": "lifecycle", "state": state])
    })
  }

  private func startMemory() {
    observers.append(NotificationCenter.default.addObserver(forName: UIApplication.didReceiveMemoryWarningNotification, object: nil, queue: .main) { [weak self] _ in
      print("[FlutterSystemEvents] memory callback: state=warning level=0")
      self?.events?(["type": "memory", "state": "warning", "level": 0])
    })
  }

  private func stopAll() {
    observers.forEach(NotificationCenter.default.removeObserver)
    observers.removeAll()
    pathMonitor?.cancel()
    pathMonitor = nil
    currentNetworkMonitor?.cancel()
    currentNetworkMonitor = nil
    stopBattery()
    stopOrientation()
  }

  private func startNetwork() {
    let monitor = NWPathMonitor()
    monitor.pathUpdateHandler = { [weak self] path in
      DispatchQueue.main.async {
        self?.events?(networkEvent(from: path))
      }
    }
    pathMonitor = monitor
    monitor.start(queue: DispatchQueue.global(qos: .utility))
  }

  private func currentNetwork(_ result: @escaping FlutterResult) {
    currentNetworkMonitor?.cancel()
    let monitor = NWPathMonitor()
    currentNetworkMonitor = monitor
    var completed = false
    monitor.pathUpdateHandler = { [weak self] path in
      DispatchQueue.main.async {
        guard !completed else { return }
        completed = true
        result(networkEvent(from: path))
        monitor.cancel()
        if self?.currentNetworkMonitor === monitor {
          self?.currentNetworkMonitor = nil
        }
      }
    }
    monitor.start(queue: DispatchQueue.global(qos: .utility))
  }

  private func startBattery() {
    previousBatteryMonitoring = UIDevice.current.isBatteryMonitoringEnabled
    UIDevice.current.isBatteryMonitoringEnabled = true
    observers.append(NotificationCenter.default.addObserver(forName: UIDevice.batteryLevelDidChangeNotification, object: nil, queue: .main) { [weak self] _ in
      self?.emitBattery()
    })
    observers.append(NotificationCenter.default.addObserver(forName: UIDevice.batteryStateDidChangeNotification, object: nil, queue: .main) { [weak self] _ in
      self?.emitBattery()
    })
    emitBattery()
  }

  private func emitBattery() {
    let device = UIDevice.current
    let state: String
    switch device.batteryState {
    case .charging:
      state = "charging"
    case .full:
      state = "full"
    case .unplugged:
      state = "discharging"
    case .unknown:
      state = "unknown"
    @unknown default:
      state = "unknown"
    }
    let level = device.batteryLevel >= 0 ? Int(device.batteryLevel * 100) : -1
    events?(["type": "battery", "level": level, "charging": state == "charging" || state == "full", "state": state])
  }

  private func stopBattery() {
    if let previousBatteryMonitoring {
      UIDevice.current.isBatteryMonitoringEnabled = previousBatteryMonitoring
      self.previousBatteryMonitoring = nil
    }
  }

  private func startOrientation() {
    previousOrientationNotifications = UIDevice.current.isGeneratingDeviceOrientationNotifications
    UIDevice.current.beginGeneratingDeviceOrientationNotifications()
    observers.append(NotificationCenter.default.addObserver(forName: UIDevice.orientationDidChangeNotification, object: nil, queue: .main) { [weak self] _ in
      self?.emitOrientation()
    })
    emitOrientation()
  }

  private func emitOrientation() {
    let orientation = orientationName(from: UIDevice.current.orientation)
    if orientation == lastOrientation { return }
    lastOrientation = orientation
    events?(["type": "orientation", "orientation": orientation])
  }

  private func stopOrientation() {
    if previousOrientationNotifications == false {
      UIDevice.current.endGeneratingDeviceOrientationNotifications()
    }
    previousOrientationNotifications = nil
    lastOrientation = nil
  }

  private func startTime() {
    observeTime(UIApplication.significantTimeChangeNotification, reason: "timeChanged")
    observeTime(NSNotification.Name.NSSystemTimeZoneDidChange, reason: "timezoneChanged")
    observeTime(NSNotification.Name.NSCalendarDayChanged, reason: "dateChanged")
  }

  private func observeTime(_ name: Notification.Name, reason: String) {
    observers.append(NotificationCenter.default.addObserver(forName: name, object: nil, queue: .main) { [weak self] _ in
      self?.events?(["type": "time", "reason": reason])
    })
  }

  private func startScreen() {
    observers.append(NotificationCenter.default.addObserver(forName: UIScreen.brightnessDidChangeNotification, object: nil, queue: .main) { [weak self] _ in
      self?.events?(["type": "screen", "change": "brightness", "brightness": UIScreen.main.brightness])
    })
    observers.append(NotificationCenter.default.addObserver(forName: UIApplication.protectedDataDidBecomeAvailableNotification, object: nil, queue: .main) { [weak self] _ in
      self?.events?(["type": "screen", "change": "unlocked"])
    })
  }

  private struct EventConfig {
    let keyboard: Bool
    let lifecycle: Bool
    let network: Bool
    let memory: Bool
    let battery: Bool
    let orientation: Bool
    let time: Bool
    let screen: Bool

    static let legacy = EventConfig(keyboard: true, lifecycle: true, network: true, memory: true, battery: false, orientation: true, time: true, screen: true)

    static func from(_ arguments: Any?) -> EventConfig {
      guard let map = arguments as? [String: Any] else { return legacy }
      return EventConfig(
        keyboard: map["keyboard"] as? Bool == true,
        lifecycle: map["lifecycle"] as? Bool == true,
        network: map["network"] as? Bool == true,
        memory: map["memory"] as? Bool == true,
        battery: map["battery"] as? Bool == true,
        orientation: map["orientation"] as? Bool == true,
        time: map["time"] as? Bool == true,
        screen: map["screen"] as? Bool == true
      )
    }
  }
}

func orientationName(from orientation: UIDeviceOrientation) -> String {
  switch orientation {
  case .portrait:
    return "portraitUp"
  case .portraitUpsideDown:
    return "portraitDown"
  case .landscapeLeft:
    return "landscapeLeft"
  case .landscapeRight:
    return "landscapeRight"
  default:
    return "unknown"
  }
}

func networkEvent(from path: NWPath) -> [String: Any] {
  let networkType: String
  if path.status != .satisfied {
    networkType = "none"
  } else if path.usesInterfaceType(.wifi) {
    networkType = "wifi"
  } else if path.usesInterfaceType(.cellular) {
    networkType = "cellular"
  } else if path.usesInterfaceType(.wiredEthernet) {
    networkType = "ethernet"
  } else {
    networkType = "other"
  }
  return ["type": "network", "online": path.status == .satisfied, "networkType": networkType]
}
